org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java

```diff
@@ -0,0 +1,311 @@
+        return fileUri;
+      }
+    }
+
+    if (fqn == null) {
+      return fileUri;
+    }
+
+    IType outerClass;
+    IJavaElement iMember;
+    try {
+      outerClass = javaProject.findType(fqn);
+
+      if (outerClass == null) {
+        return fileUri;
+      }
+
+      String source;
+      if (outerClass.isBinary()) {
+        IClassFile classFile = outerClass.getClassFile();
+        source = classFile.getSource();
+      } else {
+        ICompilationUnit unit = outerClass.getCompilationUnit();
+        source = unit.getSource();
+      }
+
+      Document document = new Document(source);
+      IRegion region = document.getLineInformation(lineNumber);
+      int start = region.getOffset();
+      int end = start + region.getLength();
+
+      iMember = binSearch(outerClass, start, end);
+    } catch (JavaModelException | BadLocationException e) {
+      throw new IllegalArgumentException(e);
+    }
+
+    if (iMember instanceof IType) {
+      return ((IType) iMember).getFullyQualifiedName();
+    }
+
+    if (iMember != null) {
+      fqn = ((IMember) iMember).getDeclaringType().getFullyQualifiedName();
+
+      // figure out if the fqn represents a local inner class
+      while (iMember != null && !(iMember instanceof CompilationUnit)) {
+        if (iMember instanceof SourceType
+            && !((SourceType) iMember).isAnonymous()
+            && iMember.getParent() instanceof SourceMethod) {
+
+          fqn = fqn.replace("$" + iMember.getElementName(), "$1" + iMember.getElementName());
+        }
+
+        iMember = iMember.getParent();
+      }
+
+      return fqn;
+    }
+
+    throw new IllegalArgumentException(
+        format("Unable to find source code for '%s' in the project '%s'", fileUri, javaProject));
+  }
+
+  /**
+   * Finds resources by a fqn.
+   *
+   * @param params contains fqn
+   * @return all resources are identified by the given fqn
+   */
+  public static List<ResourceLocationParameters> findResourcesByFqn(
+      List<Object> params, IProgressMonitor pm) {
+    Preconditions.checkArgument(params.size() >= 1, "FQN is expected.");
+
+    String fqn = (String) params.get(0);
+
+    if (pm.isCanceled()) {
+      throw new OperationCanceledException();
+    }
+
+    Pair<char[][], char[][]> fqnPair = prepareFqnToSearch(fqn);
+    List<IType> types;
+    try {
+      types =
+          findTypeByFqn(
+              fqnPair.getKey(), fqnPair.getValue(), SearchEngine.createWorkspaceScope(), pm);
+    } catch (JavaModelException e) {
+      throw new RuntimeException(e);
+    }
+
+    if (types.isEmpty()) {
+      throw new RuntimeException("Type with fully qualified name: " + fqn + " was not found");
+    }
+
+    IType type = types.get(0); // TODO we need handle few result! It's temporary solution.
+    if (type.isBinary()) {
+      IClassFile classFile = type.getClassFile();
+      int libId = classFile.getAncestor(IPackageFragmentRoot.PACKAGE_FRAGMENT_ROOT).hashCode();
```

Comment:
We REALLY should centralize production of id's for java elements. I realize this needs to be done outside the scope of this PR.